Google.Cloud.Functions.Invoker/DependencyInjection/FunctionsStartupAttribute.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright 2020, Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace Google.Cloud.Functions.Invoker.DependencyInjection
+{
+    /// <summary>
+    /// Attribute used to declare <see cref="FunctionsStartup"/> types, primarily used to
+    /// configure dependency injection within functions. When multiple types are specified,
+    /// the order in which they are used is undefined.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Assembly, Inherited = false, AllowMultiple = true)]
+    public sealed class FunctionsStartupAttribute : Attribute
+    {
+        /// <summary>
+        /// The startup type to use for additional configuration.
+        /// </summary>
+        public Type StartupType { get; set; }
+
+        /// <summary>
+        /// Constructs a new instance.
+        /// </summary>
+        /// <param name="startupType">The Type of the <see cref="FunctionsStartup"/> class to register.</param>
+        public FunctionsStartupAttribute(Type startupType) => StartupType = startupType;
```

Comment:
nit: Could you maybe check that it's really a FunctionStartup derived type?

Reply:
It's probably a bad idea to do that *in the attribute constructor* given that it gets called by the deepest darkest bits of CLR magic. The stack trace would probably be horrible.
I suspect the exception from the `Cast` call later would be clearer.

Reply:
Current stack trace:

```text
Unhandled exception. System.InvalidCastException: Unable to cast object of type 'Google.Cloud.Functions.SampleHttpFunction.NotStartup' to type 'Google.Cloud.Functions.Invoker.DependencyInjection.FunctionsStartup'.
   at System.Linq.Enumerable.CastIterator[TResult](IEnumerable source)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Builder.CreateStartups() in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 198
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Builder.Build() in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 141
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Create(Assembly functionAssembly, String[] commandLine, ConfigurationVariableProvider variableProvider) in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 69
   at Google.Cloud.Functions.Invoker.EntryPoint.StartAsync(Assembly functionAssembly, String[] args) in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\EntryPoint.cs:line 62
   at AutoGeneratedProgram.<Main>(String[] args)
```

Exception with a check in the attribute constructor:

```text
Unhandled exception. System.ArgumentException: Google.Cloud.Functions.SampleHttpFunction.NotStartup doesn't derive from FunctionsStartup
   at Google.Cloud.Functions.Invoker.DependencyInjection.FunctionsStartupAttribute..ctor(Type startupType) in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\DependencyInjection\Fun
ctionsStartupAttribute.cs:line 36
   at System.Reflection.CustomAttribute._CreateCaObject(RuntimeModule pModule, RuntimeType type, IRuntimeMethodInfo pCtor, Byte** ppBlob, Byte* pEndBlob, Int32* pcNamedArgs)
   at System.Reflection.CustomAttribute.CreateCaObject(RuntimeModule module, RuntimeType type, IRuntimeMethodInfo ctor, IntPtr& blob, IntPtr blobEnd, Int32& namedArgs)
   at System.Reflection.CustomAttribute.AddCustomAttributes(ListBuilder`1& attributes, RuntimeModule decoratedModule, Int32 decoratedMetadataToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilde
r`1 derivedAttributes)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeAssembly assembly, RuntimeType caType)
   at System.Reflection.RuntimeAssembly.GetCustomAttributes(Type attributeType, Boolean inherit)
   at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType, Boolean inherit)
   at System.Reflection.CustomAttributeExtensions.GetCustomAttributes[T](Assembly element)
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Builder.CreateStartups() in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 198
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Builder.Build() in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 141
   at Google.Cloud.Functions.Invoker.FunctionEnvironment.Create(Assembly functionAssembly, String[] commandLine, ConfigurationVariableProvider variableProvider) in C:\users\jon\Test\Projects\functions-framework-dotnet\
Google.Cloud.Functions.Invoker\FunctionEnvironment.cs:line 69
   at Google.Cloud.Functions.Invoker.EntryPoint.StartAsync(Assembly functionAssembly, String[] args) in C:\users\jon\Test\Projects\functions-framework-dotnet\Google.Cloud.Functions.Invoker\EntryPoint.cs:line 62
   at AutoGeneratedProgram.<Main>(String[] args)
```

So not *terrible* - but I think it's okay where it is.

Reply:
Yep, agree!